### PR TITLE
Add the multi-task support and a simple example

### PR DESCRIPTION
**Pull request by lyttonhao**
Replace PR #1382 
@tqchen

### CODE REVIEW

File: example/multi-task/example_multi_task.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    @property
+    def provide_data(self):
+        return self.data_iter.provide_data
+
+    @property
+    def provide_label(self):
+        provide_label = self.data_iter.provide_label[0]
+        # Different labels should be used here for actual application
+        return [('softmax1_label', provide_label[1]), \
+                ('softmax2_label', provide_label[1])]
```
Comment (futurely) [inline example/multi-task/example_multi_task.py]: How to preprocess an actual multi-label dataset into proper format and write a data iterator for it?

**Comment by tqchen** [comment]
This seems to be great. It would be nice if you can write a bit more in the markdown, on how the data iterator works and what is the objective function.

After that, please use http://mxnet.readthedocs.org/en/latest/contribute.html#git-workflow-howtos to merge with master and we can merge it in.

Thanks for the contribution!

**Comment by lyttonhao** [comment]
@tqchen I have updated the README.md file now.

**Comment by tqchen** [comment]
Sorry for delayed action. This is now merged. thanks!

**Comment by haria** [comment]
Hi @lyttonhao 

I tried your code on my own dataset, I create train.rec val.rec with label_width=2 option, and  use ImageRecorderIter as below
```python
def mnist_iterator(batch_size, input_shape):
    data_shape = input_shape
    train = mx.io.ImageRecordIter(
        path_imgrec = "./train.rec",
        #path_imglist="./train_train.txt",
        mean_r      = 123.68,
        mean_g      = 116.779,
        mean_b      = 103.939,
        data_shape  = data_shape,
        batch_size  = batch_size,
        rand_crop   = True,
        rand_mirror = True,
        label_width = 2
        )

    val = mx.io.ImageRecordIter(
        path_imgrec = "./val.rec",
        #path_imglist="./train_val.txt",
        mean_r      = 123.68,
        mean_g      = 116.779,
        mean_b      = 103.939,
        rand_crop   = False,
        rand_mirror = False,
        data_shape  = data_shape,
        batch_size  = batch_size,
        label_width = 2
        )
    return (train, val)
```

train_train.txt looks like below
>#image_index label1_id label2_id image_path
>24309	50	1	/female/star/User1/0.jpg
>24939	51	0	/male/star/User2/0.jpg

I change input_shape  to (3, 28, 28)

after execute python example_multi_task.py I find Train-multi-accuracy_0 and Train-multi-accuracy_1, Validation-multi-accuracy_0 and Validation-multi-accuracy_1 are exactly same, so I think label_2 is not correctly loaded, your code use label_1 always.

Do you have any updated code that can make multi-label classification  work correctly?
Thank you!

**Comment by lyttonhao** [comment]
Hi @haria, have you written your own iterator correctly by providing different labels. For example, This [line](https://github.com/dmlc/mxnet/blob/master/example/multi-task/example_multi_task.py#L55) should output two different labels.

**Comment by haria** [comment]
Hi @lyttonhao , you are right, I think I didn't provide a correct way to iter on my dataset, I change your example code a little, instead of directly duplicate label, I created a new label based on original label, that is "Whether the digit is >=5" it has two available value[0,1],  
```python
        #modify example_multi_task.py line 54
        label2 = mx.nd.array(label.asnumpy()>=5).astype('float32')
        return mx.io.DataBatch(data=batch.data, label=[label, label2], \
                pad=batch.pad, index=batch.index)
```

after this change, I run python  example_multi_task.py and I got following result.
[log.txt](https://github.com/dmlc/mxnet/files/259235/log.txt)
it seems not converge at all, can you guess the reason?
Thank you.

**Comment by lyttonhao** [comment]
@haria I think you need to add a new [`fc` layer](https://github.com/dmlc/mxnet/blob/master/example/multi-task/example_multi_task.py#L20)  to predict the second label. For example,
```
fc4 = mx.symbol.FullyConnected(data = act2, name='fc4', num_hidden=2)
sm2 = mx.symbol.SoftmaxOutput(data = fc4, name = 'softmax2')
```

**Comment by haria** [comment]
Hi @lyttonhao , thank you so much for your help, Now I am able to training on my dataset.
Thank you!

**Comment by tqchen** [comment]
@haria Do you mind to improve the example by adding a few paragraph your lessons on this? So others can benefit from what you learnt

**Comment by haria** [comment]
sure I will, I will submit a pull request this weekend.

**Comment by YuetianLiu** [comment]
Hi, @lyttonhao @haria , I have use the example of multi task code to test my datasets. Each data has two labels, so in the data iterator, label_width=2. and  it doesn't work. The error is as follows:
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: InferShape Error in softmax1's label argument
Corresponding keyword of symbol: softmax1_label
Shape inconsistent, Provided =(100,2), inferred shape=(100,)

My own iterator is as follow:
    def next(self):
        batch = self.data_iter.next()
        label = batch.label[0]
        return mx.io.DataBatch(data=batch.data, label=[label, label], \
                pad=batch.pad, index=batch.index)

and my softmax1 and softmax2 layer are as follows:
    fc1 = mx.symbol.FullyConnected( data=flatten1, num_hidden=8, name='fc1')
    softmax1 = mx.symbol.SoftmaxOutput( data=fc1, name="softmax1")
    fc2 = mx.symbol.FullyConnected( data=flatten2, num_hidden=2, name='fc2')
    softmax2 = mx.symbol.SoftmaxOutput( data=fc2, name="softmax2")
    symbol_lsit = [softmax1,softmax2]
    group = mx.symbol.Group(symbol_lsit)
    return group

Can you help me how to solve this problem?

Thank you very much!

**Comment by haria** [comment]
Hi @YuetianLiu 
Please refer to my tutorial
https://github.com/haria/mxnet-multi-task-example/blob/master/multi-task.ipynb

actually, I modified data iterator as below.
```python
class Multi_mnist_iterator(mx.io.DataIter):
    '''multi label mnist iterator'''

    def __init__(self, data_iter):
        super(Multi_mnist_iterator, self).__init__()
        self.data_iter = data_iter
        self.batch_size = self.data_iter.batch_size

    @property
    def provide_data(self):
        return self.data_iter.provide_data

    @property
    def provide_label(self):
        provide_label = self.data_iter.provide_label[0]
        # Different labels should be used here for actual application
        return [('softmax1_label', provide_label[1]), \
                ('softmax2_label', provide_label[1])]

    def hard_reset(self):
        self.data_iter.hard_reset()

    def reset(self):
        self.data_iter.reset()

    def next(self):
        batch = self.data_iter.next()
        label = batch.label[0] # the original label here
        label2 = mx.nd.array(label.asnumpy()>5).astype('float32') # new label based on the original label, output 0 or 1 here
        return mx.io.DataBatch(data=batch.data, label=[label, label2], \
                pad=batch.pad, index=batch.index)
```

Good Luck!

**Comment by YuetianLiu** [comment]
@haria Thanks! It is a great tutorial, I have solved my problem.

**Comment by codemarsyu** [comment]
@haria In the data iterator, 
`@property
  def provide_label(self):
        provide_label = self.data_iter.provide_label[0]
        # Different labels should be used here for actual application
        return [('softmax1_label', provide_label[1]), \
                ('softmax2_label', provide_label[1])]`
Can we just use the below, if the data_iter such as NDArray support multiple label input (list of label array). 
 `return self.data_iter.provide_label`

**Pull request merged**